man/dmdb.1
.TH "DMDB" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmdb \(en stores observations received from message queue in database
.SH SYNOPSIS
\f[B]dmdb\f[R] \-\-help
.PP
\f[B]dmdb\f[R] \-\-version
.PP
\f[B]dmdb\f[R] \-\-\f[B]database\f[R] \f[I]file\f[R] \-\-\f[B]node\f[R]
\f[I]id\f[R] [\-\-\f[B]ipc\f[R]] [\-\-\f[B]logger\f[R] \f[I]name\f[R]]
[\-\-\f[B]debug\f[R]] [\-\-\f[B]verbose\f[R]]
.PP
\f[B]dmdb\f[R] \-\-\f[B]config\f[R] \f[I]file\f[R] [\-\-\f[B]name\f[R]
\f[I]name\f[R]]
.SH DESCRIPTION
The \f[B]dmdb\f[R] program collects observations from a POSIX message
queue and stores them in an SQLite database.
The name of the message queue matches the given \f[B]dmdb\f[R] name, by
default \f[CR]dmdb\f[R].
The IPC option enables process synchronisation via POSIX semaphores.
The semaphore value is increased from 0 to 1 for every observation that
has been received.
The name of the semaphore equals the \f[B]dmdb\f[R] name.
Only a single process shall wait for the semaphore.
.SH OPTIONS
.TP
\f[B]\-\-config\f[R], \f[B]\-c\f[R] \f[I]file\f[R]
Path to configuration file.
.TP
\f[B]\-\-database\f[R], \f[B]\-d\f[R] \f[I]file\f[R]
Path to SQLite observation database.
.TP
\f[B]\-\-debug\f[R], \f[B]\-D\f[R]
Forward logs messages of level \f[CR]LL_DEBUG\f[R] via IPC (if logger is
set).
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-ipc\f[R], \f[B]\-Q\f[R]
Uses a POSIX semaphore for process synchronisation.
The name of the semaphore matches the instance name (with leading
\f[CR]/\f[R]).
The semaphore is set to 1 each time an observation has been received.
Only a single process shall wait for this semaphore, otherwise, reading
occurs in round\-robin fashion.
.TP
\f[B]\-\-logger\f[R], \f[B]\-l\f[R] \f[I]name\f[R]
Name of logger.
If set, sends logs to \f[I]dmlogger(1)\f[R] process of given name.
.TP
\f[B]\-\-name\f[R], \f[B]\-n\f[R] \f[I]name\f[R]
Name of program instance, configuration, POSIX message queue, and POSIX
semaphore (default is \f[CR]dmdb\f[R]).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Node id.
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print log messages to \f[I]stderr\f[R].
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Program executed without errors.
.TP
\f[B]1\f[R]
Failure.
Program execution failed.
.SH ENVIRONMENT
.TP
\f[B]DM_LOGGER\f[R]
Name of logger instance to send logs to.
.TP
\f[B]NO_COLOR\f[R]
Disable ANSI colour output.
.SH EXAMPLE
Create a message queue \f[CR]/dmdb\f[R], wait for incoming observations,
and store them in the given database:
.IP
.EX
$ dmdb \-n dmdb \-N dummy\-node \-d /var/dmpack/observ.db \-V
.EE
.SH SEE ALSO
\f[I]dmdbctl(1)\f[R], \f[I]dminit(1)\f[R], \f[I]dmsync(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dmdb.f90
! dmdb.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmdb
    !! The database program collects observations from a POSIX message queue and
    !! stores them in a SQLite database.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmdb'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    ! Program parameters.
    integer, parameter :: APP_DB_NSTEPS   = 500                !! Number of steps before database is optimised.
    integer, parameter :: APP_DB_TIMEOUT  = DB_TIMEOUT_DEFAULT !! SQLite 3 busy timeout in mseconds.
    logical, parameter :: APP_MQ_BLOCKING = .true.             !! Observation forwarding is blocking.

    type :: app_type
        !! Application settings.
        character(ID_LEN)          :: name     = APP_NAME !! Name of database instance and POSIX semaphore.
        character(FILE_PATH_LEN)   :: config   = ' '      !! Path to configuration file.
        character(LOGGER_NAME_LEN) :: logger   = ' '      !! Name of logger (name implies IPC).
        character(FILE_PATH_LEN)   :: database = ' '      !! Path to SQLite database file.
        character(NODE_ID_LEN)     :: node_id  = ' '      !! Node id.
        logical                    :: debug    = .false.  !! Forward debug messages via IPC.
        logical                    :: ipc      = .false.  !! Use POSIX semaphore for process synchronisation.
        logical                    :: verbose  = .false.  !! Print debug messages to stderr.
    end type app_type

    class(logger_class), pointer :: logger ! Logger object.

    integer                    :: rc     ! Return code.
    type(app_type)             :: app    ! App settings.
    type(db_type)              :: db     ! Database type.
    type(posix_mqueue_type)    :: mqueue ! POSIX message queue type.
    type(posix_sem_named_type) :: sem    ! POSIX semaphore type.

    ! Initialise DMPACK.
    call dm_init()

    ! Get command-line arguments, read options from configuration file.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Initialise logger.
    logger => dm_logger_get_default()
    call logger%configure(name    = app%logger,  & ! Name of logger process.
                          node_id = app%node_id, & ! Node id.
                          source  = app%name,    & ! Log source.
                          debug   = app%debug,   & ! Forward debug messages via IPC.
                          ipc     = .true.,      & ! Enable IPC (if logger is set).
                          verbose = app%verbose)   ! Print logs to standard error.
    call logger%info('started ' // APP_NAME)

    call init(app, db, mqueue, sem, error=rc)
    if (dm_is_error(rc)) call shutdown(rc)

    call run(app, db, mqueue, sem)
    call shutdown(E_NONE)
contains
    subroutine shutdown(error)
        !! Cleans up and stops program.
        integer, intent(in) :: error !! DMPACK error code.

        integer :: rc, stat

        stat = merge(STOP_FAILURE, STOP_SUCCESS, dm_is_error(error))

        call dm_db_close(db, error=rc)
        if (dm_is_error(rc)) call logger%error('failed to close database', error=rc)

        call dm_posix_mqueue_close(mqueue, error=rc)
        if (dm_is_error(rc)) call logger%error('failed to close mqueue /' // app%name, error=rc)

        call dm_posix_mqueue_unlink(mqueue, error=rc)
        if (dm_is_error(rc)) call logger%error('failed to unlink mqueue /' // app%name, error=rc)

        if (app%ipc) then
            call dm_posix_sem_close(sem, error=rc)
            if (dm_is_error(rc)) call logger%error('failed to close semaphore /' // app%name, error=rc)

            call dm_posix_sem_unlink(sem, error=rc)
            if (dm_is_error(rc)) call logger%error('failed to unlink semaphore /' // app%name, error=rc)
        end if

        call logger%info('stopped ' // APP_NAME, error=error)
        call dm_stop(stat)
    end subroutine shutdown

    subroutine init(app, db, mqueue, sem, error)
        !! Initialises program.
        type(app_type),             intent(inout)         :: app    !! App type.
        type(db_type),              intent(out)           :: db     !! Database type.
        type(posix_mqueue_type),    intent(out)           :: mqueue !! POSIX message queue type.
        type(posix_sem_named_type), intent(out)           :: sem    !! POSIX semaphore type.
        integer,                    intent(out), optional :: error  !! Error code.

        integer :: rc

        init_block: block
            ! Open SQLite database.
            rc = dm_db_open(db, path=app%database, timeout=APP_DB_TIMEOUT)

            if (dm_is_error(rc)) then
                call logger%error('failed to open database ' // app%database, error=rc)
                exit init_block
            end if

            if (.not. dm_db_table_has_observs(db)) then
                call logger%error('database tables not found', error=E_INVALID)
                exit init_block
            end if

            call logger%debug('opened database ' // app%database)

            ! Open observation message queue for reading.
            rc = dm_posix_mqueue_open(mqueue, type=TYPE_OBSERV, name=app%name, access=POSIX_MQUEUE_RDONLY)

            if (dm_is_error(rc)) then
                call logger%error('failed to open mqueue /' // trim(app%name) // ': ' // dm_posix_error_message(), error=rc)
                exit init_block
            end if

            call logger%debug('opened mqueue /' // app%name)

            ! Create semaphore for IPC.
            if (app%ipc) then
                rc = dm_posix_sem_open(sem, name=app%name, value=0, create=.true.)

                if (dm_is_error(rc)) then
                    call logger%error('failed to open semaphore /' // app%name, error=rc)
                    exit init_block
                end if

                call logger%debug('opened semaphore /' // app%name)
            end if

            call dm_posix_signal_register(signal_callback)
        end block init_block

        if (present(error)) error = rc
    end subroutine init

    subroutine run(app, db, mqueue, sem)
        !! Opens observation message queue for reading, and stores received
        !! derived types in database.
        type(app_type),             intent(inout) :: app    !! App settings.
        type(db_type),              intent(inout) :: db     !! Database type.
        type(posix_mqueue_type),    intent(inout) :: mqueue !! Message queue type.
        type(posix_sem_named_type), intent(inout) :: sem    !! Semaphore type.

        integer           :: rc, steps, value
        type(observ_type) :: observ

        steps = 0

        ipc_loop: do
            ! Blocking read from POSIX message queue.
            call logger%debug('waiting for observ on mqueue /' // app%name)
            rc = dm_posix_mqueue_read(mqueue, observ)

            if (dm_is_error(rc)) then
                call logger%error('failed to read from mqueue /' // trim(app%name) // ', next attempt in 30 sec', error=rc)
                call dm_sleep(30)
                cycle ipc_loop
            end if

            call logger%debug('received observ ' // observ%name)

            if (.not. dm_observ_is_valid(observ)) then
                call logger%error('invalid observ ' // trim(observ%name), error=E_INVALID)
                cycle ipc_loop
            end if

            if (dm_db_has_observ(db, observ%id)) then
                call logger%warning('observ ' // trim(observ%id) // ' exists', error=E_EXIST)
                cycle ipc_loop
            end if

            db_loop: do
                ! Add observation to database.
                rc = dm_db_insert(db, observ)

                ! Retry if database is busy.
                if (rc == E_DB_BUSY) then
                    call logger%debug('database busy', error=rc)
                    call dm_db_sleep(APP_DB_TIMEOUT)
                    cycle db_loop
                end if

                ! Handle database error.
                if (dm_is_error(rc)) then
                    call logger%error('failed to insert observ ' // observ%name, error=rc)
                    exit db_loop
                end if

                call logger%debug('inserted observ ' // observ%name)

                ! Post semaphore.
                if (app%ipc) then
                    rc = dm_posix_sem_value(sem, value)

                    if (dm_is_error(rc)) then
                        call logger%error('failed to get value of semaphore /' // app%name, error=rc)
                        exit db_loop
                    end if

                    if (value /= 0) exit db_loop

                    rc = dm_posix_sem_post(sem)

                    if (dm_is_error(rc)) then
                        call logger%error('failed to post to semaphore /' // app%name, error=rc)
                        exit db_loop
                    end if
                end if

                exit db_loop
            end do db_loop

            ! Forward observation if any receivers are left.
            rc = dm_posix_mqueue_forward(observ, name=app%name, blocking=APP_MQ_BLOCKING)

            if (steps == 0) then
                rc = dm_db_optimize(db)

                if (dm_is_error(rc)) then
                    call logger%error('failed to optimize database', error=rc)
                else
                    call logger%debug('optimized database')
                end if
            end if

            ! Increase optimise step counter.
            steps = modulo(steps + 1, APP_DB_NSTEPS)
        end do ipc_loop
    end subroutine run

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS AND CONFIGURATION FILE.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and settings from configuration file.
        type(app_type), intent(out) :: app !! App type.

        type(arg_parser_class) :: parser

        ! Required and optional command-line arguments.
        call parser%add('name',     short='n', type=ARG_TYPE_ID)       ! -n, --name <id>
        call parser%add('config',   short='c', type=ARG_TYPE_FILE)     ! -c, --config <path>
        call parser%add('logger',   short='l', type=ARG_TYPE_ID)       ! -l, --logger <id>
        call parser%add('database', short='d', type=ARG_TYPE_DATABASE) ! -d, --database <path>
        call parser%add('node',     short='N', type=ARG_TYPE_ID)       ! -N, --node <id>
        call parser%add('debug',    short='D', type=ARG_TYPE_LOGICAL)  ! -D, --debug
        call parser%add('ipc',      short='Q', type=ARG_TYPE_LOGICAL)  ! -Q, --ipc
        call parser%add('verbose',  short='V', type=ARG_TYPE_LOGICAL)  ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('name',   app%name)
        call parser%get('config', app%config)

        ! Read configuration from file.
        rc = read_config(app)
        if (dm_is_error(rc)) return

        ! Overwrite configuration.
        call parser%get('logger',   app%logger)
        call parser%get('database', app%database)
        call parser%get('node',     app%node_id)
        call parser%get('debug',    app%debug)
        call parser%get('ipc',      app%ipc)
        call parser%get('verbose',  app%verbose)

        rc = validate(app)
    end function read_args

    integer function read_config(app) result(rc)
        !! Reads configuration from (Lua) file if path is not emty.
        type(app_type), intent(inout) :: app !! App type.

        type(config_class) :: config

        rc = E_NONE
        if (.not. dm_string_has(app%config)) return

        rc = config%open(app%config, app%name)

        if (dm_is_ok(rc)) then
            call config%get('logger',   app%logger)
            call config%get('database', app%database)
            call config%get('node',     app%node_id)
            call config%get('debug',    app%debug)
            call config%get('ipc',      app%ipc)
            call config%get('verbose',  app%verbose)
        end if

        call config%close()
    end function read_config

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        if (.not. dm_id_is_valid(app%name)) then
            call dm_error_out(rc, 'invalid name')
            return
        end if

        if (dm_string_has(app%logger) .and. .not. dm_id_is_valid(app%logger)) then
            call dm_error_out(rc, 'invalid logger name')
            return
        end if

        if (.not. dm_string_has(app%database)) then
            call dm_error_out(rc, 'missing database')
            return
        end if

        if (.not. dm_file_exists(app%database)) then
            call dm_error_out(rc, 'database ' // trim(app%database) // ' does not exist')
            return
        end if

        if (.not. dm_id_is_valid(app%node_id)) then
            call dm_error_out(rc, 'invalid or missing node id')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine signal_callback(signum) bind(c)
        !! C-interoperable signal handler that closes database, removes message
        !! queue, and stops program.
        integer(c_int), intent(in), value :: signum

        call logger%debug('exit on on signal ' // dm_posix_signal_name(signum))
        call shutdown(E_NONE)
    end subroutine signal_callback

    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a, 1x, a)', dm_lua_version(.true.), dm_db_version(.true.)
    end subroutine version_callback
end program dmdb
